File: solutions/python/wordy/3/wordy.py
import re


def answer(question):
    question = question.removeprefix("What is").removesuffix("?").strip()
    if not question:
        raise ValueError("syntax error")

    pattern = r"-?\d+|plus|minus|multiplied by|divided by"
    tokens = re.findall(pattern, question)

    remaining = re.sub(pattern, "", question).strip()
    if remaining:
        raise ValueError("unknown operation")

    if not tokens or not re.match(r"-?\d+$", tokens[0]):
        raise ValueError("syntax error")
    result = int(tokens[0])

    i = 1
    while i < len(tokens):
        op = tokens[i]
        if i + 1 >= len(tokens) or not re.match(r"-?\d+$", tokens[i + 1]):
            raise ValueError("syntax error")
        num = int(tokens[i + 1])
        if op == "plus":
            result += num
        elif op == "minus":
            result -= num
        elif op == "multiplied by":
            result *= num
        elif op == "divided by":
            result //= num
        else:
            raise ValueError("unknown operation")
        i += 2

    return result

File: solutions/ruby/wordy/1/wordy.rb
class WordProblem
  OPERATORS = {
    'plus' => :+, 'minus' => :-,
    'multiplied by' => :*, 'divided by' => :/
  }

  def initialize(question)
    @question = question
  end

  def answer
    tokens = tokenize
    raise ArgumentError if tokens.empty? || tokens.length.even?
    result = tokens.shift
    until tokens.empty?
      op = tokens.shift
      raise ArgumentError unless op.is_a?(Symbol)
      num = tokens.shift
      raise ArgumentError unless num.is_a?(Integer)
      result = result.send(op, num)
    end
    result
  end

  private

  def tokenize
    m = @question.match(/^What is (.+)\?$/)
    raise ArgumentError unless m
    expr = m[1]
    tokens = []
    i = 0
    while i < expr.length
      if expr[i] == ' '
        i += 1
        next
      end
      if expr[i] =~ /-|\d/
        num = expr[i..].match(/^-?\d+/)
        raise ArgumentError unless num
        tokens << num[0].to_i
        i += num[0].length
      else
        op = OPERATORS.keys.find { |k| expr[i..].start_with?(k) }
        raise ArgumentError unless op
        tokens << OPERATORS[op]
        i += op.length
      end
    end
    tokens
  end
end
